### PR TITLE
Dynamic Dashboard: Reviews card states and unit tests

### LINKED ISSUE #12742: Last 3 reviews card logic

### PR DESCRIPTION
**Pull request by hafizrahman**
Please do not merge until target is trunk

Closes: #12742

## Description

This PR completes the Reviews card task by adding the following:
1. Finalizing UI for loading state,
2. UI for error state,
3. UI for empty state (both for when there's no reviews at all in the site, and when there's no reviews after filtering). The wording here follows [Android's implementation](https://github.com/woocommerce/woocommerce-android/pull/11491/files#diff-b931a16277ac6133750e3d68c243da133842255a53e451802acd40aad3645e81R423-R424).
4. Dismiss and onError retry trackings
5. Unit tests

## Testing instructions

Please check and compare with screenshots posted below:

1. Start with no internet, ensure "error state" is shown.
2. Retry, ensure this is tracked: `Tracked dynamic_dashboard_card_retry_tapped, ...` with `type: reviews` included.
3. Restart test, now with internet on a site with no reviews. ensure "Empty state case: absolutely no reviews" case shown.
4. Add one review on the site, but do not approve it. Go back to site, filter by "Approved". Ensure "Empty state case: no filtered reviews" case is shown.
5. Try reloading and ensure loading state is shown correctly.
6. HIde the card, ensure this is tracked: `Tracked dynamic_dashboard_hide_card_tapped, ...` with `type: reviews` included.

### Error case:
  

| Empty state case: absolutely no reviews | Empty state case:  no filtered reviews |
|-|-|
|Simulator Screenshot - iPhone 15 iOS 17 4 - 2024-05-22 at 09 42 07 | Simulator Screenshot - iPhone 15 iOS 17 4 - 2024-05-22 at 09 39 49 |

| Loading state case: no reviews ever loaded | Loading state case: some reviews were shown before |
|-|-|
| Simulator Screenshot - iPhone 15 iOS 17 4 - 2024-05-20 at 16 54 32 | Simulator Screenshot - iPhone 15 iOS 17 4 - 2024-05-20 at 16 54 21  |

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">This PR is tagged with <code>status: do not merge</code> label(s).</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WooCommerce/Classes/ViewRelated/Dashboard/Reviews/ReviewsDashboardCard.swift
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Dashboard/Reviews/ReviewsDashboardCard.swift]: Nit: this view is simple so can be part of this file, not necessary to be separate unless it's to be reused.

File: WooCommerce/Classes/ViewRelated/Dashboard/Reviews/ReviewsDashboardEmptyView.swift
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Dashboard/Reviews/ReviewsDashboardEmptyView.swift]: Nit: if this view is kept in `ReviewDashboardCard`, you can use the current filter to display in the message. Something like: "No reviews matching Hold status". This would be more explicit and concise.

File: WooCommerce/WooCommerceTests/ViewRelated/Dashboard/Reviews/ReviewsDashboardCardViewModelTests.swift
Comment (itsmeichigo) [inline WooCommerce/WooCommerceTests/ViewRelated/Dashboard/Reviews/ReviewsDashboardCardViewModelTests.swift]: We should compare the data rather than just the count, to make sure that the details are correct.

Comment (hafizrahman) [reply WooCommerce/WooCommerceTests/ViewRelated/Dashboard/Reviews/ReviewsDashboardCardViewModelTests.swift]: Thanks, updated in https://github.com/woocommerce/woocommerce-ios/pull/12806/commits/8739e01865f91f53a138a6f3aa2bd12f9de5597f

File: WooCommerce/Classes/ViewRelated/Dashboard/Reviews/ReviewsDashboardCard.swift
Comment (hafizrahman) [reply WooCommerce/Classes/ViewRelated/Dashboard/Reviews/ReviewsDashboardCard.swift]: good point, updated in https://github.com/woocommerce/woocommerce-ios/pull/12806/commits/811f143498d9f08ec28e692a662dbdb5edc614e4

File: WooCommerce/Classes/ViewRelated/Dashboard/Reviews/ReviewsDashboardEmptyView.swift
Comment (hafizrahman) [reply WooCommerce/Classes/ViewRelated/Dashboard/Reviews/ReviewsDashboardEmptyView.swift]: Yup, that's more useful. Added in https://github.com/woocommerce/woocommerce-ios/pull/12806/commits/811f143498d9f08ec28e692a662dbdb5edc614e4

**Review by itsmeichigo — COMMENTED** [review]
Hey @hafizrahman 👋  As discussed in the previous PR, I think we should not always show the review list when there is existing data. From my testing, when refreshing the card, we show the list without the shimmering animation, making it confusing to understand if data is being reloaded.

**Comment by hafizrahman** [comment]
@itsmeichigo I think something might have gone wrong / unaccounted for after merging from the latest branch, I was just responding to the previous feedbacks without adding any logic change. Let me check, thanks!

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/206">18.9</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by hafizrahman** [comment]
@itsmeichigo I decided to revamp the loading state to match similar behavior in InboxDashboardCard and have placeholder data displayed always when loading. I also simplified the syncing to only show items when there's full text with product names and notifications loaded, just makes things easier to follow that way.

Added in [b758898](https://github.com/woocommerce/woocommerce-ios/pull/12806/commits/b75889894c7bcb7391d9b0be8c8e485a4741e316) and [8d87f7f](https://github.com/woocommerce/woocommerce-ios/pull/12806/commits/8d87f7fdce4d7752557ff4fcdc9854cc113bf181)

Video of latest loading state:


https://github.com/woocommerce/woocommerce-ios/assets/266376/7128474e-3004-4fb8-aed8-37e0611249d9

**Review by itsmeichigo — APPROVED** [review]
Thanks Hafiz, LGTM :shipit:

**Pull request merged**
